src/display_manager.py

```
@@ -164,9 +166,24 @@ def _setup_matrix(self):
+                f"Matrix initialization failed — running in fallback/simulation mode "
+                f"(size {fallback_width}x{fallback_height}). Error: {e}. "
+                "On Raspberry Pi 5: ensure rpi-rgb-led-matrix was built from the latest "
+                "submodule (re-run first_time_install.sh) and that gpio_slowdown is 4 or higher."
+            )
             # Do not raise here; allow fallback mode so web preview and non-hardware environments work
 
+        # Write hardware status file so the web UI can surface init failures
+        try:
+            import json as _json
+            _hw_status = {"ok": self.matrix is not None, "error": _init_error_str}
+            _status_path = "/tmp/led_matrix_hw_status.json"  # nosec B108
```

Ruff (0.15.12)

[error] Probable insecure usage of temporary file or directory: "/tmp/led_matrix_hw_status.json"

(S108)